### PR TITLE
[Wallet] Laggard wallet-related backports from btc 0.15

### PR DESCRIPTION
**Pull request by random-zebra**
I think these are all the remaining Bitcoin Core v0.15 PRs in the wallet area that we don't have yet, and are useful to us.
I've grouped them here since they are all pretty small, simple, and narrow-focused (on the wallet/rpcwallet files).

This changeset is based on top of
- [x] #2337

as it keeps going with the multiwallet implementation, adding:
- RPC endpoint support
- `listwallets` RPC
- wallet name in `getwalletinfo` RPC
- `wallet_multiwallet.py` functional test

As in some areas we are much closer to upstream, some of the commits needed adaptations (especially the functional tests). A couple of commits have been extended to include shield-related code.

List of upstream PRs backported/adapted:

- bitcoin#9906  : Disallow copy constructor CReserveKeys
- bitcoin#9830  : Add safe flag to listunspent result
- bitcoin#9993  : Initialize nRelockTime
- bitcoin#10108 : ApproximateBestSubset should take inputs by reference, not value
- bitcoin#10075 : Remove unused C++ code not covered by unit tests
- bitcoin#10257 : Add test for getmemoryinfo
- bitcoin#10295 : Move some WalletModel functions into CWallet
- bitcoin#10500 : Avoid CWalletTx copies in GetAddressBalances and GetAddressGroupings
- bitcoin#10522 : Remove unused variables
- bitcoin#10816 : Properly forbid -salvagewallet and -zapwallettxes for multi wallet
- bitcoin#10849 : Multiwallet: simplest endpoint support
- bitcoin#9380  : Separate different uses of minimum fees (only dustRelayFee)
- bitcoin#10817 : Redefine Dust and add a discard_rate (only first commit)
- bitcoin#10883 : Rename -usewallet to -rpcwallet
- bitcoin#10604 : Add listwallets RPC, include wallet name in getwalletinfo + tests
- bitcoin#10893 : Avoid running multiwallet.py twice
- bitcoin#10870 : Use wallet 0 in rpc console if running with multiple wallets
- bitcoin#10931 : Fix misleading "method not found" multiwallet errors
- bitcoin#10885 : Reject invalid wallets
- bitcoin#11022 : Basic keypool topup
- bitcoin#10976 : [MOVEONLY] Move some static functions out of wallet.h/cpp
- bitcoin#11310 : Test listwallets RPC
- bitcoin#10858 : "errors" to getblockchaininfo + unify "errors" field in get*info RPCs
- bitcoin#11492 : Fix leak in CDB constructor
- bitcoin#11476 : Avoid opening copied wallet databases simultaneously
- bitcoin#11590 : always show help-line of wallet encryption calls
- bitcoin#11289 : Add wallet backup text to import* and add* RPCs

### CODE REVIEW

**Comment by random-zebra** [comment]
Rebased on master, now that  #2337 has been merged.

**Comment by Fuzzbawls** [comment]
quick thing before doing a more complete review:

I see you've moved a lot of the tier two `.cpp` files out of the wallet library and into the common library for autotools builds. would be a good idea to do the same for cmake builds as well to maintain concurrency between the two:
```diff
Index: CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	(revision 9a868f8579704920271d0c259a28478c46704041)
+++ b/CMakeLists.txt	(date 1621481390924)
@@ -251,24 +251,10 @@
 endif()
 
 set(WALLET_SOURCES
-        ./src/activemasternode.cpp
         ./src/bip38.cpp
         ./src/wallet/db.cpp
         ./src/addressbook.cpp
         ./src/crypter.cpp
-        ./src/budget/budgetdb.cpp
-        ./src/budget/budgetmanager.cpp
-        ./src/budget/budgetproposal.cpp
-        ./src/budget/budgetvote.cpp
-        ./src/budget/finalizedbudget.cpp
-        ./src/budget/finalizedbudgetvote.cpp
-        ./src/masternode.cpp
-        ./src/masternode-payments.cpp
-        ./src/masternode-sync.cpp
-        ./src/tiertwo_networksync.cpp
-        ./src/masternodeconfig.cpp
-        ./src/masternodeman.cpp
-        ./src/messagesigner.cpp
         ./src/zpiv/mintpool.cpp
         ./src/wallet/hdchain.cpp
         ./src/wallet/rpcdump.cpp
@@ -355,8 +341,15 @@
 target_include_directories(ZEROCOIN_A PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/src)
 
 set(COMMON_SOURCES
+        ./src/activemasternode.cpp
         ./src/base58.cpp
         ./src/bip38.cpp
+        ./src/budget/budgetdb.cpp
+        ./src/budget/budgetmanager.cpp
+        ./src/budget/budgetproposal.cpp
+        ./src/budget/budgetvote.cpp
+        ./src/budget/finalizedbudget.cpp
+        ./src/budget/finalizedbudgetvote.cpp
         ./src/consensus/params.cpp
         ./src/consensus/upgrades.cpp
         ./src/chainparams.cpp
@@ -379,6 +372,12 @@
         ./src/invalid.cpp
         ./src/key.cpp
         ./src/keystore.cpp
+        ./src/masternode.cpp
+        ./src/masternode-payments.cpp
+        ./src/masternode-sync.cpp
+        ./src/masternodeconfig.cpp
+        ./src/masternodeman.cpp
+        ./src/messagesigner.cpp
         ./src/netaddress.cpp
         ./src/netbase.cpp
         ./src/policy/feerate.cpp
@@ -393,6 +392,7 @@
         ./src/script/script_error.cpp
         ./src/spork.cpp
         ./src/sporkdb.cpp
+        ./src/tiertwo_networksync.cpp
         ./src/warnings.cpp
         )
 add_library(COMMON_A STATIC ${BitcoinHeaders} ${COMMON_SOURCES})
```

**Comment by random-zebra** [comment]
Good idea @Fuzzbawls :+1: . Patch squashed into 49c201ef78b139d5b60bb9f4f00e048e4f68fbdd

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2459,9 +2461,9 @@ void CWallet::GetAvailableP2CSCoins(std::vector<COutput>& vCoins) const {
 /**
  * Test if the transaction is spendable.
  */
-static bool CheckTXAvailabilityInternal(const CWalletTx* pcoin, bool fOnlyConfirmed, int& nDepth)
+static bool CheckTXAvailabilityInternal(const CWalletTx* pcoin, bool fOnlySafe, int& nDepth)
 {
-    if (fOnlyConfirmed && !pcoin->IsTrusted()) return false;
+    if (fOnlySafe && !pcoin->IsTrusted()) return false;
```
Comment (furszy) [inline src/wallet/wallet.cpp:2378]: We could add a return param for the `IsTrusted` result like we are doing with the depth. So the function isn't called twice to set the safe field of the returned output.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3004,6 +3004,20 @@ std::map<std::pair<CTxDestination, Optional<CTxDestination>>, std::vector<COutpu
...
+    std::vector<SaplingNoteEntry> notes;
+    Optional<libzcash::SaplingPaymentAddress> dummy = nullopt;
+    m_sspk_man->GetFilteredNotes(notes, dummy);
+    for (const auto& note : notes) {
+        result[note.address].emplace_back(std::move(note));
+    }
+
+    return result;
+}
+
```
Comment (furszy) [inline src/wallet/wallet.cpp:2932]: nit: would be better to place this inside the sspkm.

**Review by furszy — APPROVED** [review]
Code review ACK 1cb5cb2114d5d9cad9a0adac0a8dd593ec691b3e.
Left two small topic and a nit that can be done in a subsequent PR.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (random-zebra) [reply src/wallet/wallet.cpp]: Moved it in the SaplingScriptPubKeyMan and made a wrapper in the wallet.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (random-zebra) [reply src/wallet/wallet.cpp]: Yeah. `IsTrusted` should be extremely fast for the majority of txes (locktime=0), but have added it anyway.

File: src/miner.cpp [hunk 3]
```diff
@@ -118,9 +118,11 @@ void BitcoinMiner(CWallet* pwallet, bool fProofOfStake)
...
 
     // Each thread has its own key and counter
-    Optional<CReserveKey> opReservekey{nullopt};
-    if (!fProofOfStake) {
-        opReservekey = CReserveKey(pwallet);
+    CReserveKey reservekey(pwallet);
+    CReserveKey* pReservekey = &reservekey;
+    if (fProofOfStake) {
+        pReservekey->ReturnKey();
+        pReservekey = nullptr;
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/miner.cpp]: I have switched to a smart `unique_ptr` instead, so don't  have to worry about the scope of the referenced object.

**Review by furszy — APPROVED** [review]
utACK c710b030fcfd11aec3b0c86fcd095bf5e4139237 after rebase and covered feedback.

**Comment by random-zebra** [comment]
Rebased on master, fixing conflicts.

**Review by furszy — APPROVED** [review]
re-utACK e840ab7 after rebase.

**Review by furszy — APPROVED** [review]
re-re-utACK d5526bd.

**Review by Fuzzbawls — APPROVED** [review]
ACK d5526bdd67784c3a6064e8709f4db84b62b7f76f

**Comment by random-zebra** [comment]
Merging...

**Pull request merged**